src/CustomElementInternals.js

```diff
@@ -38,6 +155,159 @@ export default class CustomElementInternals {
+     * A map from `localName`s of definitions that were defined *after* the last
+     * flush to unupgraded elements matching that definition, in document order.
+     * @type {!Map<string, !Array<!HTMLElement>>}
+     */
+    const elementsWithPendingDefinitions = new Map();
+    for (let i = 0; i < pendingDefinitions.length; i++) {
+      elementsWithPendingDefinitions.set(pendingDefinitions[i].localName, []);
+    }
+
+    this.patchAndUpgradeTree(document, {
+      upgrade: element => {
+        // Ignore the element if it has already upgraded or failed to upgrade.
+        if (element.__CE_state !== undefined) return;
+
+        const localName = element.localName;
+
+        // If there is an applicable pending definition for the element, add the
+        // element to the list of elements to be upgraded with that definition.
+        const pendingElements = elementsWithPendingDefinitions.get(localName);
+        if (pendingElements) {
+          pendingElements.push(element);
+        // If there is *any other* applicable definition for the element, add it
+        // to the list of elements with stable definitions that need to be upgraded.
+        } else if (this.localNameToDefinition(localName) ||
+            this.localNameToLazyDefinition(localName)) {
+          elementsWithStableDefinitions.push(element);
+        }
+      },
+    });
+
+    // Upgrade elements with 'stable' definitions first.
+    for (let i = 0; i < elementsWithStableDefinitions.length; i++) {
+      this.upgradeElement(elementsWithStableDefinitions[i]);
+    }
+
+    // Upgrade elements with 'pending' definitions in the order they were defined.
+    while (pendingDefinitions.length > 0) {
+      const definition = pendingDefinitions.shift();
+      const localName = definition.localName;
+
+      // Attempt to upgrade all applicable elements.
+      const pendingUpgradableElements = elementsWithPendingDefinitions.get(definition.localName);
+      for (let i = 0; i < pendingUpgradableElements.length; i++) {
+        this.upgradeElement(pendingUpgradableElements[i]);
+      }
+
+      // Resolve any promises created by `whenDefined` for the definition.
+      const deferred = this._whenDefinedDeferred.get(localName);
+      if (deferred) {
+        deferred.resolve(undefined);
+      }
+    }
+  }
+
+  /**
+   *
+   * @param {string} localName
+   */
+  flushLazyDefinition(localName) {
+    const pendingLazyDefinition = this.localNameToLazyDefinition(localName);
+
+    // only process this pending definition the first time and not while it's pending.
+    if (!pendingLazyDefinition || pendingLazyDefinition === PendingLazyDefinitionMarker) {
+      return;
+    }
+
+    // mark this definition in a pending state.
+    this.setLazyDefinition(localName, PendingLazyDefinitionMarker);
+
+    let constructorOrPromise;
+    try {
+      constructorOrPromise = pendingLazyDefinition.constructorGetter();
+    } catch (e) {}
+    if (!constructorOrPromise) {
+      return;
```

Comment:
Maybe we should type check and throw?

Reply:
Actually, I'm on the fence about throwing here. `connectTree` / `disconnectTree` don't currently catch errors even though [they're really supposed to be reported](https://html.spec.whatwg.org/multipage/custom-elements.html#invoke-custom-element-reactions).